File: fun/aoc2022/day11.py
from collections import defaultdict
from math import lcm

example_data = """Monkey 0:
  Starting items: 79, 98
  Operation: new = old * 19
  Test: divisible by 23
    If true: throw to monkey 2
    If false: throw to monkey 3

Monkey 1:
  Starting items: 54, 65, 75, 74
  Operation: new = old + 6
  Test: divisible by 19
    If true: throw to monkey 2
    If false: throw to monkey 0

Monkey 2:
  Starting items: 79, 60, 97
  Operation: new = old * old
  Test: divisible by 13
    If true: throw to monkey 1
    If false: throw to monkey 3

Monkey 3:
  Starting items: 74
  Operation: new = old + 3
  Test: divisible by 17
    If true: throw to monkey 0
    If false: throw to monkey 1
"""

INPUT_FILE_PATH = "fun/aoc2022/input_files/day11_gap.txt"


def read_file():
    with open(INPUT_FILE_PATH, "r") as f:
        return f.read().strip()


class Monkey:
    def __init__(self, input_data, divisor=3):
        self.input_data = input_data
        self.parse_input()
        self.inspect_count = 0
        self.divisor = divisor

    def parse_input(self):
        for line in self.input_data.strip().split("\n"):
            line: str = line.strip()
            if line.startswith("Monkey"):
                n = int(line.split(" ")[1].split(":")[0])
                self.n = n
            elif line.startswith("Starting"):
                items = line.split(": ")[1]
                self.items = list(map(int, items.split(",")))
            elif line.startswith("Operation"):
                op = line.split("=")[1]
                self.operation = lambda old: eval(op)
            elif line.startswith("Test"):
                self.divisible = int(line.split("by")[1])
            elif line.startswith("If true"):
                self.if_true = int(line.split("monkey")[1])
            elif line.startswith("If false"):
                self.if_false = int(line.split("monkey")[1])

    def inspect(self):
        new_items = []
        for i, item in enumerate(self.items):
            self.inspect_count += 1
            worry_level = self.operation(item) // self.divisor
            new_items.append(worry_level)
        self.items = new_items

    def throw(self):
        throw = defaultdict(list)
        for item in self.items:
            if item % self.divisible == 0:
                throw[self.if_true].append(item)
            else:
                throw[self.if_false].append(item)
        self.items = []
        return throw

    def catch(self, items: list):
        self.items += items


def solution(data: str, part=1):
    inputs = data.split("\n\n")
    monkeys = []
    for i in inputs:
        divisor = 3 if part == 1 else 1
        monkeys.append(Monkey(i, divisor=divisor))

    LCM = lcm(*[m.divisible for m in monkeys])

    N = 20 if part == 1 else 10000
    for round in range(N):
        for monkey in monkeys:
            monkey: Monkey
            monkey.inspect()
            thrown = monkey.throw()
            for m, item in thrown.items():
                monkeys[m].catch([i % LCM for i in item])

    inspect_counts = [m.inspect_count for m in monkeys]
    return sorted(inspect_counts)[-2] * sorted(inspect_counts)[-1]


if __name__ == "__main__":
    assert solution(example_data, part=1) == 10605
    print("Part 1: ", solution(read_file(), part=1))

    assert solution(example_data, part=2) == 2713310158
    print("Part 2: ", solution(read_file(), part=2))

File: fun/aoc2022/day11.jl
struct Monkey{V<:AbstractVector, S<:AbstractString, I<:Integer}
    items::V
    op1::S 
    op2::S 
    div::I
    throw_true::I
    throw_false::I
end

function solution(ifile; part)
    # read file
    monkeys = init_monkeys(ifile)

    # least common multiple of monkey divisors
    N = 1
    for m ∈ monkeys
        N *= m.div
    end

    # number of rounds
    if part == 1
        n_rounds = 20
    elseif part == 2
        n_rounds = 10000
    end

    # count inspections
    inspections = zeros(length(monkeys))
    for _=1:n_rounds
        monkeys, round_inspections = play_round(monkeys, N; part)
        inspections += round_inspections
    end

    return prod(sort(inspections, rev=true)[1:2])
end

function play_round(monkeys, N; part)
    inspections = zeros(length(monkeys))
    for i ∈ eachindex(monkeys)
        m = monkeys[i]
        for item ∈ m.items
            if part == 1
                new_item = div(op(m, item), 3)
            elseif part == 2
                new_item = mod(op(m, item), N)
            end
            if rem(new_item, m.div) == 0
                push!(monkeys[m.throw_true].items, new_item)
            else
                push!(monkeys[m.throw_false].items, new_item)
            end
            inspections[i] += 1
        end
        for _=1:length(m.items)
            pop!(m.items)
        end
    end
    return monkeys, inspections
end

function op(m::Monkey, old)
    if m.op2 == "old"
        n = old
    else
        n = parse(Int64, m.op2)
    end

    if m.op1 == "+"
        return old + n
    elseif m.op1 == "*"
        return old * n
    end
end

function init_monkeys(ifile)
    monkeys = []
    open(ifile, "r") do f
        while !eof(f)
            # mokey # line
            readline(f)

            # get items
            line = split(readline(f))
            items = parse.(Int64, chopsuffix.(line[3:end], ","))

            # get operation
            line = split(readline(f))
            op1 = line[5] # + or *
            op2 = line[6] # number of `old`

            # get divisor
            line = split(readline(f))
            div = parse(Int64, line[4])

            # get monkey to throw to if true
            line = split(readline(f))
            throw_true = parse(Int64, line[6]) + 1

            # get monkey to throw to if false
            line = split(readline(f))
            throw_false = parse(Int64, line[6]) + 1

            # add to list
            push!(monkeys, Monkey(items, op1, op2, div, throw_true, throw_false))

            # read blank line
            readline(f)
        end
    end
    return monkeys
end

println(solution("input_files/day11_hgp.txt", part=1)) # 405 μs
println(solution("input_files/day11_hgp.txt", part=2)) # 206 ms